src/EVM/SymExec.hs

```diff
@@ -726,7 +718,48 @@ equivalenceCheck' solvers branchesA branchesB opts = do
         -- if we cannot statically determine whether or not the end states
         -- differ, then we ask the solver if the end states can differ if both
         -- sets of path conditions are satisfiable
-        _ -> Just . Set.fromList $ differingResults : extractProps aEnd <> extractProps bEnd
+        _ -> Just . Set.fromList $ resultsDiffer aEnd bEnd : extractProps aEnd <> extractProps bEnd
+
+    resultsDiffer :: Expr End -> Expr End -> Prop
+    resultsDiffer aEnd bEnd = case (aEnd, bEnd) of
+      (Success _ _ aOut aState, Success _ _ bOut bState) ->
+        case (aOut == bOut, aState == bState) of
+          (True, True) -> PBool False
+          (False, True) -> aOut ./= bOut
+          (True, False) -> statesDiffer aState bState
+          (False, False) -> statesDiffer aState bState .|| aOut ./= bOut
+      (Failure _ _ (Revert a), Failure _ _ (Revert b)) -> if a == b then PBool False else a ./= b
+      (Failure _ _ a, Failure _ _ b) -> if a == b then PBool False else PBool True
+      -- partial end states can't be compared to actual end states, so we always ignore them
+      (Partial {}, _) -> PBool False
+      (_, Partial {}) -> PBool False
+      (ITE _ _ _, _) -> internalError "Expressions must be flattened"
+      (_, ITE _ _ _) -> internalError "Expressions must be flattened"
+      (a, b) -> if a == b
+                then PBool False
+                else PBool True
+
+    statesDiffer :: Map (Expr EAddr) (Expr EContract) -> Map (Expr EAddr) (Expr EContract) -> Prop
+    statesDiffer aState bState
+      = if Set.fromList (Map.keys aState) /= Set.fromList (Map.keys bState)
```

Comment:
Just a note that when we do equivalence checking for Act we will want to check for equivalence up to a renaming of addresses. We can integrate this into this equivalence checker or write an other one for Act.